docs: TypeScript support in Custom Applications and release `v21.8` (#2624)
commercetools/merchant-center-application-kit: website/src/content/development/adding-typescript.mdx

---
title: Adding TypeScript
---

[TypeScript](https://www.typescriptlang.org/) is a typed superset of JavaScript that compiles to plain JavaScript.

Custom Applications have full support for developing applications in TypeScript.

<Info>

> This feature is available from version `>= 21.8.0`.

Reviewer: ```suggestion
This feature is available from version `21.8.0`.
```

Using the word "from" should imply greater than or equal to the stated version.

Reviewer 2: Thanks for the suggestion. If we change this, there are also other places in the docs that would need to be updated.

Can you confirm that it's clear that the "equal" part is implied? Not every user is a native english speaker, not I'm not sure if that's really obvious... 🤔

To be on the safe side, would it still be ok with you if we keep the `>=`?

Reviewer: Sure, you could keep `>=` for now, or you could consider the following as well:

```suggestion
This feature is available from version `21.8.0` onwards.
```

Reviewer 2: Thanks, I'll use the `onwards` version.

---
title: Adding TypeScript
---

[TypeScript](https://www.typescriptlang.org/) is a typed superset of JavaScript that compiles to plain JavaScript.

Custom Applications have full support for developing applications in TypeScript.

<Info>

This feature is available from version `>= 21.8.0`.

</Info>

To start a new Custom Application project with TypeScript, you can use the TypeScript starter template:

```console noPromptLines="2-3"
npx @commercetools-frontend/create-mc-app@latest \
  my-new-custom-application-project \
  --template starter-typescript
```

The TypeScript starter template is the same as the standard JS starter template in terms of functionality but it includes the additional TypeScript setup.

> # Configuration

Reviewer 3: Here we explain about the needed configuration but, at least for me, it seems like this is something that I would need to do when creating a new Custom Applications.

Perhaps we could add a note at the end of the section explaining all of these steps are already preconfigured in our starter template and users don't need to worry about setting them up.

Reviewer 2: Thanks, updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2624/commits/785d879b8aa392d7235f38076fef2b6d3e5adbee

* Install the `typescript` dependency.
* Add a `tsconfig.json` file. See [configuration](#configuration) for the tooling setup.
* Migrate the JavaScript files to TypeScript, either `.ts` or `.tsx` (if a file contains JSX syntax).

  <Info>

  Migrating source files to TypeScript can be done incrementally. We recommend starting from the "leaf files" and work your way up to the application entry point.
  The [migrating from JavaScript documentation](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html) can help with setting up a basic plan.

  All UI Kit and ApplicationKit packages provide declaration files and export useful types when necessary.

  During the migration you might need to loosen up the types strictness, for example by allowing explicit `any` types.

  ```json title="tsconfig.json"
  {
    "extends": "@commercetools-frontend/application-config/tsconfig-mc-app.json",
    "compilerOptions": {
      "noExplicitAny": false,
      "strict": false
    }
  }
  ```

  Please remember to revert that once the migration is over, as using `any` should be avoided.

  </Info>

> * Add a script command `"typecheck": "tsc --noEmit"` in your `package.json` to compile the application.

Reviewer 3: For me this is a little confusing since the command is called `typecheck`, which I understand as a command to validate the types in the application but we say this is to compile your application.

I thought our `mc-scripts build` command would also support building the TS application and the new `typecheck` command is only for "lintin" purposes.

Reviewer 2: Good point. Technically "compiling" is what TypeScript does. In fact, running `tsc` does compile the project but the option `--noEmit` does not emit any files, which makes it look like it only checks the types.

To be honest this is confusing yes, a more explicit command would have been better, something like `tsc check` or something.

Let's see if we can try rewording this:

```suggestion
* Add a script command `"typecheck": "tsc --noEmit"` in your `package.json` to run the type checks of the application.
```
